src/pudl/transform/ferc1.py

```diff
@@ -853,6 +853,533 @@ def strip_col_of_non_ints(self, col: pd.Series) -> pd.DataFrame:
+        exceptions = [
+            "hydro plants: licensed proj. no.",
+            "hydro license no.",
+            "hydro: license no.",
+            "hydro plants: licensed proj no.",
+        ]
+        not_header_if_cols_na = [
+            "construction_year",
+            "net_generation_mwh",
+            "total_cost_of_plant",
+            "capex_per_mw",
+            "opex_total",
+            "opex_fuel",
+            "opex_maintenance",
+            "fuel_cost_per_mmbtu",
+        ]
+
+        def label_header_rows(df: pd.DataFrame) -> pd.DataFrame:
+            """Label header rows.
+
+            This function labels rows it believes are possible headers based on whether
+            they contain information in certain key columns. Of those possible headers,
+            ones that contain a specific key word or phrase are dubbed headers. Leftover
+            possible header rows are evaluated as possible note rows in the
+            _label_notes_rows() function.
+            """
+            logger.info(f"{self.table_id.value}: Labeling header rows")
+
+            # Label possible header rows (based on the nan cols specified above)
+            df.loc[
+                df.filter(not_header_if_cols_na).isna().all(1), "possible_header"
+            ] = True
+
+            # Label good header rows (based on whether they contain key strings)
+            possible_header = df["possible_header"]
+            good_header = df["plant_name_ferc1"].str.contains("|".join(header_strings))
+            not_bad = ~df["plant_name_ferc1"].str.contains("|".join(nonheader_strings))
+
+            df.loc[possible_header & good_header & not_bad, "row_type"] = "header"
+            df.loc[df["plant_name_ferc1"].isin(exceptions), "row_type"] = "header"
+
+            return df
+
+        def label_total_rows(df: pd.DataFrame) -> pd.DataFrame:
+            """Label total rows."""
+            logger.info(f"{self.table_id.value}: Labeling total rows")
+
+            df.loc[df["plant_name_ferc1"].str.contains("total"), "row_type"] = "total"
+
+            # Now deal with some outliers:
+
+            # There are a couple rows that contain the phrase "amounts are for" in the
+            # plant name that contain total pertaining to total records above. This
+            # section of code moves the total row values that are reported as notes to
+            # correct row above and nulls the values in the notes row.
+            num_cols = [
+                x
+                for x in df.select_dtypes(include=["float", "Int64"]).columns.tolist()
+                if x not in ["utility_id_ferc1", "report_year", "license_id_ferc1"]
+            ]
+            bad_row = df[
+                (df["plant_name_ferc1"].str.contains("amounts are for"))
+                & (df["capacity_mw"] > 0)
+            ]
+
+            df.loc[bad_row.index, num_cols] = df.loc[bad_row.index - 1][num_cols].values
+            df.loc[bad_row.index - 1, num_cols] = np.nan
```

Comment:
Oh ya this is a little spot fix for some annoying rows. Technically they are totals, but the content is in the wrong spot (one row below). This could be a stand-alone function, but I lumped it in because it pertains to the totals. Maybe I should remove it?

Also, these rows are _technically_ totals, but not in the sense that the information is getting double counted. Rather, the wind turbine information is reported in bulk. I might spend some time trying to disentangle these rows from the "total" designation.

Reply:
Ok this is fixed. I moved all this content to a new function called `spot_fix_rows`